From: Christoph Hellwig <hch@lst.de>
Date: Tue, 9 Apr 2024 16:37:26 +0200
Subject: scsi: block: Add a helper to cancel atomic queue limit updates
Git-commit: 293066264fb45df4290897c22e69833bea5fe171
Patch-mainline: v6.10-rc1
References: bsc#1228325 jsc#PED-9783

Drivers might have to perform complex actions to determine queue limits,
and those might fail.  Add a helper to cancel a queue limit update that can
be called in those cases.

[lduncan: applied with fuzz=2 and refreshed]

Signed-off-by: Christoph Hellwig <hch@lst.de>
Link: https://lore.kernel.org/r/20240409143748.980206-2-hch@lst.de
Reviewed-by: Kanchan Joshi <joshi.k@samsung.com>
Reviewed-by: Damien Le Moal <dlemoal@kernel.org>
Reviewed-by: Hannes Reinecke <hare@suse.de>
Reviewed-by: John Garry <john.g.garry@oracle.com>
Reviewed-by: Johannes Thumshirn <johannes.thumshirn@wdc.com>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
Acked-by: Lee Duncan <lduncan@suse.com>
---
 include/linux/blkdev.h |   13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -907,6 +907,19 @@ queue_limits_start_update(struct request
 int queue_limits_commit_update(struct request_queue *q,
 		struct queue_limits *lim);
 
+/**
+ * queue_limits_cancel_update - cancel an atomic update of queue limits
+ * @q:		queue to update
+ *
+ * This functions cancels an atomic update of the queue limits started by
+ * queue_limits_start_update() and should be used when an error occurs after
+ * starting update.
+ */
+static inline void queue_limits_cancel_update(struct request_queue *q)
+{
+	mutex_unlock(&q->limits_lock);
+}
+
 /*
  * Access functions for manipulating queue properties
  */
